tests/python/pants_test/engine/test_scheduler.py

```diff
@@ -357,3 +386,13 @@ def test_trace_includes_rule_exception_traceback(self):
                                raise Exception(f'An exception for {type(x).__name__}')
                            Exception: An exception for B''').lstrip() + '\n\n',  # Traces include two empty lines after.
                                trace)
+
+
+class RuleIndexingErrorTest(TestBase):
+
+  def test_non_frozen_dataclass_error(self):
+
+    with self.assertRaisesWithMessage(TypeError, dedent("""\
+      type check error in class RootRule: 1 error type checking constructor arguments:
+      field 'output_type' was invalid: type <class 'pants_test.engine.test_scheduler.NonFrozenDataclass'> is a dataclass declared without `frozen=True`! The engine requires that fields in params are immutable for stable hashing!""")):
+      RootRule(NonFrozenDataclass)
```

Comment:
Would be great to add a test ensuring `frozen_after_init` works properly.

Thanks for adding these tests! Very helpful to test the intersection of V2 engine and dataclasses.

Reply:
Added testing!